test/Toolbox/Util/GraphSpec.hs

```diff
@@ -0,0 +1,86 @@
+module Toolbox.Util.GraphSpec (spec) where
+
+import qualified Data.List as L
+import Test.Hspec
+  ( Spec,
+    describe,
+    it,
+    shouldBe,
+  )
+import Toolbox.Channel (ModuleGraphInfo (..))
+import Toolbox.Util.Graph
+  ( ClusterState (..),
+    ClusterVertex (..),
+    degreeInvariant,
+    filterOutSmallNodes,
+    fullStep,
+    getBiDepGraph,
+    makeSeedState,
+    totalNumberInvariant,
+  )
+
+testGraph :: [(Int, [Int])]
+testGraph =
+  [ (1, [])
+  , (2, [1, 4, 5, 6])
+  , (3, [6])
+  , (4, [])
+  , (5, [4, 7, 8])
+  , (6, [8])
+  , (7, [9, 10])
+  , (8, [9])
+  , (9, [])
+  , (10, [])
+  ]
+
+testGraphInfo :: ModuleGraphInfo
+testGraphInfo =
+  ModuleGraphInfo
+    { mginfoModuleNameMap =
+        [ (1, "A")
+        , (2, "B")
+        , (3, "C")
+        , (4, "D")
+        , (5, "E")
+        , (6, "F")
+        , (7, "G")
+        , (8, "H")
+        , (9, "I")
+        , (10, "J")
+        ]
+    , mginfoModuleDep = testGraph
+    , mginfoModuleTopSorted = []
+    }
+
+spec :: Spec
+spec =
+  describe "Toolbox.Util.Graph greedy clustering" $ do
+    let bgr = getBiDepGraph testGraphInfo
+        allNodes = fmap fst $ mginfoModuleNameMap testGraphInfo
+        nNodes = length allNodes
+        largeNodes = filterOutSmallNodes testGraphInfo
+        smallNodes = allNodes L.\\ largeNodes
+        seeds =
+          ClusterState
+            { clusterStateClustered =
+                fmap (\i -> (Cluster i, [i])) largeNodes
+            , clusterStateUnclustered = smallNodes
+            }
+        clustering0 = seeds
+        graph0 = makeSeedState largeNodes bgr
+        r0 = (clustering0, graph0)
+        r1@(clustering1, graph1) = fullStep r0
+        _r2@(clustering2, graph2) = fullStep r1
+
+    it "should preserve the number of all the nodes" $ do
+      totalNumberInvariant clustering0 `shouldBe` nNodes
+      totalNumberInvariant clustering1 `shouldBe` nNodes
+      totalNumberInvariant clustering2 `shouldBe` nNodes
+
+    it "should keep in-degree being matched with out-degree of the derived bi-dep graph after clustering" $ do
+      let (outdeg0, indeg0) = degreeInvariant graph0
+          (outdeg1, indeg1) = degreeInvariant graph1
+          (outdeg2, indeg2) = degreeInvariant graph2
+      outdeg0 `shouldBe` indeg0
+      outdeg1 `shouldBe` indeg1
+      outdeg2 `shouldBe` indeg2
```

Comment:
these assertions sound like good candidates for property testing? i don't think i'll ever endorse property testing for mercury-web-backend, but this is a smaller break-out library, so it might be appropriate here.

Reply:
definitely agree. generation of graphs needs some work, so i will add property tests later as we gather more testable properties for a few graph algorithms. thanks for suggestion.